File: behavioral/dart/memento_structured/text_editor.dart
class TextEditorMemento {
  final String content;
  final String timestamp;
  
  TextEditorMemento(this.content) : timestamp = DateTime.now().toString();
  
  String get getContent => content;
  String get getTimestamp => timestamp;
}

class TextEditor {
  String _content = '';
  
  void write(String text) {
    _content += text;
    print('Text: "$text" added');
  }
  
  void setContent(String newContent) {
    _content = newContent;
  }
  
  String get content => _content;
  
  TextEditorMemento save() {
    print('Saving current state...');
    return TextEditorMemento(_content);
  }
  
  void restore(TextEditorMemento memento) {
    _content = memento.getContent;
    print('Restored state from ${memento.getTimestamp}');
    print('Current content: "$_content"');
  }
}

class History {
  final List<TextEditorMemento> _mementos = [];
  
  void push(TextEditorMemento memento) {
    _mementos.add(memento);
  }
  
  TextEditorMemento? pop() {
    if (_mementos.isEmpty) {
      return null;
    }
    return _mementos.removeLast();
  }
  
  int get size => _mementos.length;
}

File: behavioral/typescript/memento_structured/text_editor.ts
export class TextEditorMemento {
  private content: string;
  private timestamp: string;
  
  constructor(content: string) {
    this.content = content;
    this.timestamp = this.getCurrentTime();
  }
  
  getContent(): string {
    return this.content;
  }
  
  getTimestamp(): string {
    return this.timestamp;
  }
  
  private getCurrentTime(): string {
    return new Date().toLocaleString();
  }
}

export class TextEditor {
  private content: string = '';
  
  write(text: string): void {
    this.content += text;
    console.log(`Text: "${text}" added`);
  }
  
  setContent(newContent: string): void {
    this.content = newContent;
  }
  
  getContent(): string {
    return this.content;
  }
  
  save(): TextEditorMemento {
    console.log('Saving current state...');
    return new TextEditorMemento(this.content);
  }
  
  restore(memento: TextEditorMemento): void {
    this.content = memento.getContent();
    console.log(`Restored state from ${memento.getTimestamp()}`);
    console.log(`Current content: "${this.content}"`);
  }
}

export class History {
  private mementos: TextEditorMemento[] = [];
  
  push(memento: TextEditorMemento): void {
    this.mementos.push(memento);
  }
  
  pop(): TextEditorMemento | null {
    if (this.mementos.length === 0) {
      return null;
    }
    return this.mementos.pop()!;
  }
  
  size(): number {
    return this.mementos.length;
  }
}
